File: backend/gym-v2/src/main/java/com/gym/v2/core/response/PageResponse.java
package com.gym.v2.core.response;

import java.util.List;
import org.springframework.data.domain.Page;

/**
 * Spring Data Page nesnesini API katmanından izole etmek için kullanılan generic zarf
 * yapısı.
 *
 * @param <T> İçerik tipi
 */
public record PageResponse<T>(List<T> content, int pageNumber, int pageSize, long totalElements, int totalPages,
		boolean isLast) {
	public static <T> PageResponse<T> from(Page<T> page) {
		return new PageResponse<>(page.getContent(), page.getNumber(), page.getSize(), page.getTotalElements(),
				page.getTotalPages(), page.isLast());
	}
}

File: frontend/gymapp-v2/lib/core/network/page_response.dart
class PageResponse<T> {
  final List<T> content;
  final int page;
  final int size;
  final int totalElements;
  final int totalPages;

  PageResponse({
    required this.content,
    required this.page,
    required this.size,
    required this.totalElements,
    required this.totalPages,
  });

  factory PageResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Object? json) fromJsonT,
  ) {
    return PageResponse<T>(
      content: (json['content'] as List<dynamic>).map(fromJsonT).toList(),
      page: json['page'] as int? ?? 0,
      size: json['size'] as int? ?? 0,
      totalElements: json['totalElements'] as int? ?? 0,
      totalPages: json['totalPages'] as int? ?? 0,
    );
  }

  bool get isLast => page >= totalPages - 1;
}
